### ISSUE TITLE
Closing of drop-downs of each attribute is user unfriendly

### ISSUE THREAD

**Issue by dumblob** [nicolaskruchten/pivottable#673]
Currently drop-down menus (opened by clicking on the small triangle) do not close in the following cases:

a) when one clicks on the triangle
b) when one clicks somewhere outside of the drop-down menu
c) when one opens another drop-down (this is especially frustrating because of overlapping and "stacking" upon each other, etc.)

I find the current behavior really unintuitive and user unfriendly. Could you please consider implementing closing of drop-down menus in all the situations above?

**Comment by nicolaskruchten** [comment]
I have considered all of these cases in the past and settled on the current behaviour to give the user explicit control over whether they want to apply or cancel whatever modifications they might have made to the filter while it was open. The original behaviour was to auto-close-and-apply when you clicked outside of it, and it caused usability problems.

**Comment by dumblob** [comment]
>The original behaviour was to auto-close-and-apply when you clicked outside of it, and it caused usability problems.

That sounds actually very usable to me. What exactly was the usability problem?

If we really want to leave things how they are right now, we must totally reconsider positioning of drop-downs. They must never overlap each other, but also not the pivat table itself. So in case a first drop-down menu is opened, this would first shift the pivot table down by the high of the drop-down menu. The second opened drop-down menu would be positioned next to the first opened menu (to not overlap it), the third menu next to the second one and so on. First after closing all drop-down menus the pivot table would move back upwards by the height of a drop-down menu.

**Comment by nicolaskruchten** [comment]
You're coming on a bit strong here with "we must totally reconsider" etc :)

I've been maintaining this library and receiving input for a few years now and it's been working fine for the majority of folks so I'm frankly not all that inclined to make major changes...

**Comment by dumblob** [comment]
>You're coming on a bit strong here with "we must totally reconsider" etc :)

Yep, that's just my attempt to counterbalance the experience I had with drop-downs.

>it's been working fine for the majority of folks

That's good news, no doubt. But it doesn't say anything about their satisfaction with the UI :wink:.

Anyway, I'm really interested in all the specific reasons for deliberately skipping every single feature out of those three I proposed in the initial comment. I'd like to thoroughly understand the workflows others employ to get me to the same level of knowledge (and consequently to the level of discussion) as you have. For me the current drop-downs behavior is a major drawback.

**Comment by nicolaskruchten** [comment]
Let me see if I can try to rephrase: if the dropdown auto-closes, there has to be a default action, either apply or cancel. Neither of those is a non-surprising default, in that there are always some users who will expect auto-apply and some who will not. The choice I have made, which I remain comfortable with, is to not choose a default action and force the user to explicitly apply or cancel via the provided buttons.

**Comment by nicolaskruchten** [comment]
See for example this feedback: #154 against the idea of auto-apply on close.

That said, there has also been echoes of your current feedback re multiple dropdowns open at the same time: #622

**Comment by powelleric** [comment]
I was actually really wishing for the ability to auto-close the drop-down when clicking outside of it as well.  I do appreciate that that can be a little ambiguous though in terms of whether the modifications are applied.
On the bright side, I'm happy to learn that the functionality is possible in theory.  It would be really terrific if that could be an option that the developer enables as it seems different people have different ideas of what is intuitive/user-friendly in their context.

**Comment by d-aven** [comment]
I agree that filter modals are not ideal. In my opinion, default behaviour should be auto-close-dontApply and it should be given the possibility to manange.

Another problem is in showing modal out of screen
![image](https://user-images.githubusercontent.com/19648531/35220066-a0e73c34-ff7d-11e7-839b-3dcdd638bfe0.png)
In this case user is not able to apply or cancel.

**Comment by nicolaskruchten** [comment]
I've implemented some different behaviour in the sibling library https://react-pivottable.js.org/ ... instantaneous application of checkboxes, no cancellation, and 'tear away' functionality to place the dropdowns wherever you like. What do people think of these features? Which ones would be worth back-porting to this library?

**Comment by dumblob** [comment]
> What do people think of these features?

I've tried it right now and they look handy, but they're not much much useful compared to making filter dialogs/windows auto-close. Actually this "reactive" solution would be perfect combined with auto-close menus.
